metaverse_components/pet.js

```diff
@@ -90,45 +141,106 @@ export default (app, component) => {
       _unwear();
     }
   });
-  
+
   const smoothVelocity = new THREE.Vector3();
   const lastLookQuaternion = new THREE.Quaternion();
   const _getAppDistance = () => {
-    const position = localVector.copy(player.position);
-    position.y = 0;
+    const position = player.position.clone();
+
+    // setting the height of the y as the height of the ground
+    const collisionToGround = physics.raycast(position, downQuaternion);
+    if (collisionToGround) {
+      position.y = collisionToGround.point[1];
+    }
+
     const distance = app.position.distanceTo(position);
-    return distance;
+    return { distance, position };
   };
   const minDistance = 1;
-  const _isFar = distance => (distance - minDistance) > 0.01;
-  useFrame(({timestamp, timeDiff}) => {
+  const _isFar = (distance) => distance - minDistance > 0.01;
+
+  let fallLoopStartTimeS = 0;
+  let lastGravityH = 0;
+  const startFallingPosition = new THREE.Vector3(0, 0, 0);
+
+  useFrame(({ timestamp, timeDiff }) => {
+    const nowS = timestamp / 1000;
+    const timeDiffS = timeDiff / 1000;
+    const timeDiffCapped = Math.min(Math.max(timeDiff, 0), 100);
+    const timeDiffSCapped = timeDiffCapped / 1000;
+
+    if (fallLoopStartTimeS == 0) {
+      fallLoopStartTimeS = nowS;
+      startFallingPosition.copy(app.position);
+      localVector.set(0, 0, 0);
+    }
+    const t = nowS - fallLoopStartTimeS;
+    const h = 0.5 * physicsScene.getGravity().y * t * t;
+
+    const displacement = localVector.set(0, h - lastGravityH, 0);
+    smoothVelocity.add(displacement);
+    const flags = physicsScene.moveCharacterController(

```

Comment:
This seems a duplication of character movement code, so should be charged with npc characters.